File: pc/ygoapi/ygodata.pas
unit ygodata;

{$mode objfpc}{$H+}

interface

uses
  Classes, SysUtils, fpjson, jsonparser, jsonscanner, fgl, nativeapi, ygorequest;

type
  TPackageInfo = class
  public
    season: string;
    url: string;
    name: string;
    abbr: string;
  end;

  TLimitInfo = class
  public
    limit: integer;
    color: string;
    hashid: string;
    name: string;
  end;

  TCardPackInfo = class
  public
    url: string;
    name: string;
    date: string;
    abbr: string;
    rare: string;
  end;

  TCardPackInfoList = specialize TFPGList<TCardPackInfo>;

  { TCardDetail }

  TCardDetail = class
  public
    name: string;
    japname: string;
    enname: string;
    cardtype: string;
    password: string;
    limit: string;
    belongs: string;
    rare: string;
    pack: string;
    effect: string;
    race: string;
    element: string;
    level: string;
    atk: string;
    def: string;
    link: string;
    linkArrow: string;
    packs: TCardPackInfoList;
    adjust: string;
    wiki: string;
    constructor Create;
    destructor Destroy; override;
  end;

  TCardInfo = class
  public
    cardid: integer;
    hashid: string;
    name: string;
    japname: string;
    enname: string;
    cardtype: string;
  end;

  TCardInfoList = specialize TFPGList<TCardInfo>;


  { TSearchResult }

  TSearchResult = class
  public
    data: TCardInfoList;
    page: integer;
    pageCount: integer;
    constructor Create;
    destructor Destroy; override;
  end;

  THotCard = class
  public
    hashid: string;
    name: string;
  end;

  THotPack = class
  public
    packid: string;
    name: string;
  end;

  THotCardList = specialize TFPGList<THotCard>;
  THotPackList = specialize TFPGList<THotPack>;

  { THotest }

  THotest = class
  public
    search: TStringList;
    card: THotCardList;
    pack: THotPackList;
    constructor Create;
    destructor Destroy; override;
  end;

  { TYGOData }

  TLimitList = specialize TFPGList<TLimitInfo>;
  TPackageList = specialize TFPGList<TPackageInfo>;

  TYGOData = class
  private
    class function replaceChars(str: string): string;
    class function parseSearchResult(jsonString: string): TSearchResult;
  public
    class function searchCommon(akey: string; apage: integer): TSearchResult;
    class function searchComplex(aname: string; ajapname: string; aenname: string; arace: string; aelement: string;
      aatk: string; adef: string; alevel: string; apendulum: string; alink: string; alinkarrow: string; acardtype: string;
      acardtype2: string; aeffect: string; apage: integer): TSearchResult;
    class function cardDetail(hashid: string): TCardDetail;
    class function limit(): TLimitList;
    class function packageList(): TPackageList;
    class function packageDetail(aurl: string): TSearchResult;
    class function hostest(): THotest;
  end;


implementation

{ THotest }

constructor THotest.Create;
begin
  search := TStringList.Create;
  card := THotCardList.Create;
  pack := THotPackList.Create;
end;

destructor THotest.Destroy;
var
  i: integer;
begin
  search.Free;
  for i := card.Count - 1 downto 0 do
    card[i].Free;
  card.Free;
  for i := pack.Count - 1 downto 0 do
    pack[i].Free;
  pack.Free;
  inherited Destroy;
end;

{ TYGOData }

class function TYGOData.replaceChars(str: string): string;
begin
  str := str.Replace('&quot;', '"');
  str := str.Replace('&#039;', '''');
  str := str.Replace('　', '');
  Exit(str);
end;

class function TYGOData.parseSearchResult(jsonString: string): TSearchResult;
var
  parser: TJSONParser;
  json: TJSONObject;
  jarr: TJSONArray;
  jobj: TJSONObject;
  i: integer;
  info: TCardInfo;
begin
  Result := TSearchResult.Create;
  try
    parser := TJSONParser.Create(jsonString, []);
    json := TJSONObject(parser.Parse);
    if (json.Integers['result'] = 0) then
    begin
      Result.page := json.Integers['page'];
      Result.pageCount := json.Integers['pagecount'];
      jarr := json.Arrays['data'];
      for i := 0 to jarr.Count - 1 do
      begin
        jobj := jarr.Objects[i];
        info := TCardInfo.Create;
        info.cardid := jobj.Integers['id'];
        info.hashid := jobj.Strings['hashid'];
        info.name := replaceChars(jobj.Strings['name']);
        info.japname := replaceChars(jobj.Strings['japname']);
        info.enname := replaceChars(jobj.Strings['enname']);
        info.cardtype := jobj.Strings['cardtype'];
        Result.Data.Add(info);
      end;
    end;
    json.Free;
    parser.Free;
  except
  end;
end;

class function TYGOData.searchCommon(akey: string; apage: integer): TSearchResult;
var
  data: string = '';
  parsed: string = '';
begin
  data := TYGORequest.search(akey, apage);
  if (data <> '') then begin
    parsed := string(parse(PChar(data), 0));
  end;
  Exit(parseSearchResult(parsed));
end;

class function TYGOData.searchComplex(aname: string; ajapname: string; aenname: string; arace: string; aelement: string;
  aatk: string; adef: string; alevel: string; apendulum: string; alink: string; alinkarrow: string; acardtype: string;
  acardtype2: string; aeffect: string; apage: integer): TSearchResult;
var
  key: string = '';
begin
  if (aname <> '') then
    key += Format(' +(name:%s)', [aname]);
  if (ajapname <> '') then
    key += Format(' +(japName:%s)', [ajapname]);
  if (aenname <> '') then
    key += Format(' +(enName:%s)', [aenname]);
  if (arace <> '') then
    key += Format(' +(race:%s)', [arace]);
  if (aelement <> '') then
    key += Format(' +(element:%s)', [aelement]);
  if (aatk <> '') then
    key += Format(' +(atk:%s)', [aatk]);
  if (adef <> '') then
    key += Format(' +(def:%s)', [adef]);
  if (alevel <> '') then
    key += Format(' +(level:%s)', [alevel]);
  if (apendulum <> '') then
    key += Format(' +(pendulumL:%s)', [apendulum]);
  if (alink <> '') then
    key += Format(' +(link:%s)', [alink]);
  if (alinkarrow <> '') then
    key += Format(' +(linkArrow:%s)', [alinkarrow]);
  if (acardtype <> '') then
    key += Format(' +(cardType:%s)', [acardtype]);
  if (acardtype2 <> '') then
    key += Format(' +(cardType:%s)', [acardtype2]);
  if (aeffect <> '') then
    key += Format(' +(effect:%s)', [aeffect]);
  Exit(searchCommon(key, apage));
end;

class function TYGOData.cardDetail(hashid: string): TCardDetail;
var
  ahtml: string;
  parsed: string;
  adjust: string;
  wikiHtml: string;
  wiki: string;
  json: TJSONObject;
  jarr: TJSONArray;
  i: integer;
  parser: TJSONParser;
  obj: TJSONObject;
  pkinfo: TJSONObject;
  info: TCardPackInfo;
begin
  ahtml := TYGORequest.cardDetail(hashid);
  parsed := string(parse(PChar(ahtml), 1));
  WriteLn(parsed);
  adjust := string(parse(PChar(ahtml), 2));
  wikiHtml := TYGORequest.cardWiki(hashid);
  wiki := string(parse(PChar(wikiHtml), 3));
  parser := TJSONParser.Create(parsed, []);
  Result := TCardDetail.Create;
  try
    json := TJSONObject(parser.Parse);
    if (json.Integers['result'] = 0) then
    begin
      obj := json.Objects['data'];
      Result.name := replaceChars(obj.Strings['name']);
      Result.japname := replaceChars(obj.Strings['japname']);
      Result.enname := replaceChars(obj.Strings['enname']);
      Result.cardtype := obj.Strings['cardtype'];
      Result.password := obj.Strings['password'];
      Result.limit := obj.Strings['limit'];
      Result.belongs := obj.Strings['belongs'];
      Result.rare := obj.Strings['rare'];
      Result.pack := replaceChars(obj.Strings['pack']);
      Result.effect := replaceChars(obj.Strings['effect']);
      Result.race := obj.Strings['race'];
      Result.element := obj.Strings['element'];
      Result.level := obj.Strings['level'];
      Result.atk := obj.Strings['atk'];
      Result.def := obj.Strings['def'];
      Result.link := obj.Strings['link'];
      Result.linkArrow:= obj.Strings['linkarrow'];
      jarr := obj.Arrays['packs'];
      for i := 0 to jarr.Count - 1 do
      begin
        pkinfo := jarr.Objects[i];
        info := TCardPackInfo.Create;
        info.url := pkinfo.Strings['url'];
        info.name := replaceChars(pkinfo.Strings['name']);
        info.date := pkinfo.Strings['date'];
        info.abbr := pkinfo.Strings['abbr'];
        info.rare := pkinfo.Strings['rare'];
        Result.packs.Add(info);
      end;
      Result.adjust := adjust;
      Result.wiki := wiki;
    end;
    json.Free;
    parser.Free;
  except
  end;

end;

class function TYGOData.limit(): TLimitList;
var
  ahtml: string;
  parsed: string;
  json: TJSONObject;
  parser: TJSONParser;
  jarr: TJSONArray;
  i: integer;
  obj: TJSONObject;
  info: TLimitInfo;
begin
  ahtml := TYGORequest.limit();
  parsed := string(parse(PChar(ahtml), 4));
  Result := TLimitList.Create;
  try
    parser := TJSONParser.Create(parsed, []);
    json := TJSONObject(parser.Parse);
    if (json.Integers['result'] = 0) then
    begin
      jarr := json.Arrays['data'];
      for i := 0 to jarr.Count - 1 do
      begin
        obj := jarr.Objects[i];
        info := TLimitInfo.Create;
        info.limit := obj.Integers['limit'];
        info.color := obj.Strings['color'];
        info.hashid := obj.Strings['hashid'];
        info.name := replaceChars(obj.Strings['name']);
        Result.Add(info);
      end;
    end;
    json.Free;
    parser.Free;
  except
  end;

end;

class function TYGOData.packageList(): TPackageList;
var
  ahtml: string;
  parsed: string;
  json: TJSONObject;
  parser: TJSONParser;
  obj: TJSONObject;
  jarr: TJSONArray;
  i: integer;
  info: TPackageInfo;
begin
  ahtml := TYGORequest.packageList();
  parsed := string(parse(PChar(ahtml), 5));
  Result := TPackageList.Create;
  try
    parser := TJSONParser.Create(parsed, []);
    json := TJSONObject(parser.Parse);
    if (json.Integers['result'] = 0) then
    begin
      jarr := json.Arrays['data'];
      for i := 0 to jarr.Count - 1 do
      begin
        obj := jarr.Objects[i];
        info := TPackageInfo.Create;
        info.season := obj.Strings['season'];
        info.url := obj.Strings['url'];
        info.name := replaceChars(obj.Strings['name']);
        info.abbr := obj.Strings['abbr'];
        Result.Add(info);
      end;
    end;
    json.Free;
    parser.Free;
  except
  end;

end;

class function TYGOData.packageDetail(aurl: string): TSearchResult;
var
  ahtml: string;
  parsed: string;
begin
  ahtml := TYGORequest.packageDetail(aurl);
  parsed := string(parse(PChar(ahtml), 0));
  Exit(parseSearchResult(parsed));
end;

class function TYGOData.hostest(): THotest;
var
  ahtml: string;
  parsed: string;
  json: TJSONObject;
  parser: TJSONParser;
  jarr: TJSONArray;
  obj: TJSONObject;
  i: integer;
  infoCard: THotCard;
  infoPack: THotPack;
begin
  ahtml := TYGORequest.hotest();
  parsed := string(parse(PChar(ahtml), 6));
  Result := THotest.Create;
  try
    parser := TJSONParser.Create(parsed, []);
    json := TJSONObject(parser.Parse);

    if (json.Integers['result'] = 0) then
    begin
      jarr := json.Arrays['search'];
      for i := 0 to jarr.Count - 1 do
      begin
        Result.search.Add(jarr.Strings[i]);
      end;
      jarr := json.Arrays['card'];
      for i := 0 to jarr.Count - 1 do
      begin
        obj := jarr.Objects[i];
        infoCard := THotCard.Create;
        infoCard.hashid := obj.Strings['hashid'];
        infoCard.name := replaceChars(obj.Strings['name']);
        Result.card.Add(infoCard);
      end;
      jarr := json.Arrays['pack'];
      for i := 0 to jarr.Count - 1 do
      begin
        obj := jarr.Objects[i];
        infoPack := THotPack.Create;
        infoPack.packid := obj.Strings['packid'];
        infoPack.name := replaceChars(obj.Strings['name']);
        Result.pack.Add(infoPack);
      end;
    end;
    json.Free;
    parser.Free;
  except
  end;

end;

{ TSearchResult }

constructor TSearchResult.Create;
begin
  Data := TCardInfoList.Create;
  page:= 1;
  pageCount:= 1;
end;

destructor TSearchResult.Destroy;
var
  i: integer;
begin
  for i := Data.Count - 1 downto 0 do
    Data[i].Free;
  Data.Free;
  inherited Destroy;
end;

{ TCardDetail }

constructor TCardDetail.Create;
begin
  packs := TCardPackInfoList.Create;
end;

destructor TCardDetail.Destroy;
var
  i: integer;
begin
  for i := packs.Count - 1 downto 0 do
    packs[i].Free;
  packs.Free;
  inherited Destroy;
end;

end.

File: ios/YGOAPI/YGOAPI/YGOData.m
//
//  YGOData.m
//  YGOAPI
//
//  Created by rarnu on 2018/6/30.
//  Copyright © 2018 rarnu. All rights reserved.
//

#import "YGOData.h"
#import "YGORequest.h"
#import "YGOCAPI.h"

@implementation PackageInfo
@end
@implementation LimitInfo
@end
@implementation CardPackInfo
@end
@implementation CardDetail
@end
@implementation CardInfo
@end
@implementation SearchResult
@end
@implementation HotCard
@end
@implementation HotPack
@end
@implementation Hotest
@end

@implementation YGOData

+(NSString*) replaceChars:(NSString*) str {
    if (str == nil) {
        return @"";
    }
    str = [str stringByReplacingOccurrencesOfString:@"&quot;" withString:@"\""];
    str = [str stringByReplacingOccurrencesOfString:@"&#039;" withString:@"'"];
    str = [str stringByReplacingOccurrencesOfString:@"　" withString:@""];
    return str;
}

+(SearchResult*) parseSearchResult:(NSString*) jsonString {
    SearchResult* result = [[SearchResult alloc] init];
    @try {
        NSJSONSerialization* json = [NSJSONSerialization JSONObjectWithData:[jsonString dataUsingEncoding:NSUTF8StringEncoding] options:NSJSONReadingMutableLeaves error:nil];
        if ([[json valueForKey:@"result"] intValue] == 0) {
            result.page = [[json valueForKey:@"page"] integerValue];
            result.pageCount = [[json valueForKey:@"pagecount"] integerValue];
            NSArray* jarr = [json valueForKey:@"data"];
            NSMutableArray<CardInfo*>* arr = [NSMutableArray array];
            for (int i = 0; i < jarr.count; i++) {
                id obj = [jarr objectAtIndex:i];
                CardInfo* info = [[CardInfo alloc] init];
                info.cardid = [[obj valueForKey:@"id"] integerValue];
                info.hashid = [obj valueForKey:@"hashid"];
                info.name = [YGOData replaceChars:[obj valueForKey:@"name"]];
                info.japname = [YGOData replaceChars:[obj valueForKey:@"japname"]];
                info.enname = [YGOData replaceChars:[obj valueForKey:@"enname"]];
                info.cardtype = [obj valueForKey:@"cardtype"];
                [arr addObject:info];
            }
            result.data = arr;
        }
    } @catch (NSException *exception) {
        
    } @finally {
        
    }
    return result;
}

+(SearchResult*) searchCommon:(NSString*) key page:(NSInteger)page {
    NSString* data = [YGORequest search:key page:page];
    NSString* parsed = @"";
    if (![data isEqualToString:@""]) {
        parsed = [NSString stringWithUTF8String:parse([data UTF8String], 0)];
    }
    return [YGOData parseSearchResult:parsed];
}

+(SearchResult*) searchComplex:(NSString*) name japname:(NSString*)japname enname:(NSString*) enname race:(NSString*) race element:(NSString*) element atk:(NSString*) atk def:(NSString*) def level:(NSString*) level pendulum:(NSString*) pendulum link:(NSString*) link linkArrow:(NSString*) linkArrow cardType:(NSString*) cardType cardType2:(NSString*) cardType2 effect:(NSString*) effect page:(NSInteger)page {
    NSString* key = @"";
    if (![name isEqualToString:@""]) {
        key = [key stringByAppendingFormat:@" +(name:%@)", name];
    }
    if (![japname isEqualToString:@""]) {
        key = [key stringByAppendingFormat:@" +(japName:%@)", japname];
    }
    if (![enname isEqualToString:@""]) {
        key = [key stringByAppendingFormat:@" +(enName:%@)", enname];
    }
    if (![race isEqualToString:@""]) {
        key = [key stringByAppendingFormat:@" +(race:%@)", race];
    }
    if (![element isEqualToString:@""]) {
        key = [key stringByAppendingFormat:@" +(element:%@)", element];
    }
    if (![atk isEqualToString:@""]) {
        key = [key stringByAppendingFormat:@" +(atk:%@)", atk];
    }
    if (![def isEqualToString:@""]) {
        key = [key stringByAppendingFormat:@" +(def:%@)", def];
    }
    if (![level isEqualToString:@""]) {
        key = [key stringByAppendingFormat:@" +(level:%@)", level];
    }
    if (![pendulum isEqualToString:@""]) {
        key = [key stringByAppendingFormat:@" +(pendulumL:%@)", pendulum];
    }
    if (![link isEqualToString:@""]) {
        key = [key stringByAppendingFormat:@" +(link:%@)", link];
    }
    if (![linkArrow isEqualToString:@""]) {
        key = [key stringByAppendingFormat:@" +(linkArrow:%@)", linkArrow];
    }
    if (![cardType isEqualToString:@""]) {
        key = [key stringByAppendingFormat:@" +(cardType:%@)", cardType];
    }
    if (![cardType2 isEqualToString:@""]) {
        key = [key stringByAppendingFormat:@" +(cardType:%@)", cardType2];
    }
    if (![effect isEqualToString:@""]) {
        key = [key stringByAppendingFormat:@" +(effect:%@)", effect];
    }
    return [YGOData searchCommon:key page:page];
}

+(CardDetail*) cardDetail:(NSString*) hashid {
    NSString* data = [YGORequest cardDetail:hashid];
    NSString* wikidata = [YGORequest cardWiki:hashid];
    NSString* parsed = @"";
    NSString* adjust = @"";
    if (![data isEqualToString:@""]) {
        parsed = [NSString stringWithUTF8String:parse([data UTF8String], 1)];
        adjust = [NSString stringWithUTF8String:parse([data UTF8String], 2)];
    }
    NSString* wikiparsed = @"";
    if (![wikidata isEqualToString:@""]) {
        wikiparsed = [NSString stringWithUTF8String:parse([wikidata UTF8String], 3)];
    }
    
    CardDetail* result = [[CardDetail alloc] init];
    @try {
        NSJSONSerialization* json = [NSJSONSerialization JSONObjectWithData:[parsed dataUsingEncoding:NSUTF8StringEncoding] options:NSJSONReadingMutableLeaves error:nil];
        if ([[json valueForKey:@"result"] intValue] == 0) {
            id obj = [json valueForKey:@"data"];
            result.name = [YGOData replaceChars:[obj valueForKey:@"name"]];
            result.japname = [YGOData replaceChars:[obj valueForKey:@"japname"]];
            result.enname = [YGOData replaceChars:[obj valueForKey:@"enname"]];
            result.cardtype = [obj valueForKey:@"cardtype"];
            result.password = [obj valueForKey:@"password"];
            result.limit = [obj valueForKey:@"limit"];
            result.belongs = [obj valueForKey:@"belongs"];
            result.rare = [obj valueForKey:@"rare"];
            result.pack = [YGOData replaceChars:[obj valueForKey:@"pack"]];
            result.effect = [YGOData replaceChars:[obj valueForKey:@"effect"]];
            result.race = [obj valueForKey:@"race"];
            result.element = [obj valueForKey:@"element"];
            result.level = [obj valueForKey:@"level"];
            result.atk = [obj valueForKey:@"atk"];
            result.def = [obj valueForKey:@"def"];
            result.link = [obj valueForKey:@"link"];
            
            NSMutableArray<CardPackInfo*>* pk = [NSMutableArray array];
            NSArray* jarr = [obj valueForKey:@"packs"];
            for (int i = 0; i < jarr.count; i++) {
                id pkinfo = [jarr objectAtIndex:i];
                CardPackInfo* info = [[CardPackInfo alloc] init];
                info.url = [pkinfo valueForKey:@"url"];
                info.name = [YGOData replaceChars:[pkinfo valueForKey:@"name"]];
                info.date = [pkinfo valueForKey:@"date"];
                info.abbr = [pkinfo valueForKey:@"abbr"];
                info.rare = [pkinfo valueForKey:@"rare"];
                [pk addObject:info];
            }
            result.packs = pk;
            result.adjust = [YGOData replaceChars:adjust];
            result.wiki = [YGOData replaceChars:wikiparsed];
        }
    } @catch (NSException *exception) {
        
    } @finally {
        
    }
    return result;
}

+(NSArray<LimitInfo*>*) limit {
    NSString* data = [YGORequest limit];
    NSString* parsed = @"";
    if (![data isEqualToString:@""]) {
        parsed = [NSString stringWithUTF8String:parse([data UTF8String], 4)];
    }
    NSMutableArray<LimitInfo*>* result = [NSMutableArray array];
    @try {
        NSJSONSerialization* json = [NSJSONSerialization JSONObjectWithData:[parsed dataUsingEncoding:NSUTF8StringEncoding] options:NSJSONReadingMutableLeaves error:nil];
        if ([[json valueForKey:@"result"] intValue] == 0) {
            NSArray* jarr = [json valueForKey:@"data"];
            for (int i = 0; i < jarr.count; i++) {
                id obj = [jarr objectAtIndex:i];
                LimitInfo* info = [[LimitInfo alloc] init];
                info.limit = [[obj valueForKey:@"limit"] integerValue];
                info.color = [obj valueForKey:@"color"];
                info.hashid = [obj valueForKey:@"hashid"];
                info.name = [YGOData replaceChars:[obj valueForKey:@"name"]];
                [result addObject:info];
            }
        }
    } @catch (NSException *exception) {
        
    } @finally {
        
    }
    return result;
}

+(NSArray<PackageInfo*>*) packageList {
    NSString* data = [YGORequest packageList];
    NSString* parsed = @"";
    if (![data isEqualToString:@""]) {
        parsed = [NSString stringWithUTF8String:parse([data UTF8String], 5)];
    }
    NSMutableArray<PackageInfo*>* result = [NSMutableArray array];
    @try {
        NSJSONSerialization* json = [NSJSONSerialization JSONObjectWithData:[parsed dataUsingEncoding:NSUTF8StringEncoding] options:NSJSONReadingMutableLeaves error:nil];
        if ([[json valueForKey:@"result"] intValue] == 0) {
            NSArray* jarr = [json valueForKey:@"data"];
            for (int i = 0; i < jarr.count; i++) {
                id obj = [jarr objectAtIndex:i];
                PackageInfo* info = [[PackageInfo alloc] init];
                info.season = [obj valueForKey:@"season"];
                info.url = [obj valueForKey:@"url"];
                info.name = [YGOData replaceChars:[obj valueForKey:@"name"]];
                info.abbr = [obj valueForKey:@"abbr"];
                [result addObject:info];
            }
        }
    } @catch (NSException *exception) {
        
    } @finally {
        
    }
    return result;
}

+(SearchResult*) packageDetail:(NSString*) url {
    NSString* data = [YGORequest packageDetail:url];
    NSString* parsed = @"";
    if (![data isEqualToString:@""]) {
        parsed = [NSString stringWithUTF8String:parse([data UTF8String], 0)];
    }
    return [YGOData parseSearchResult:parsed];
}

+(Hotest*) hostest {
    NSString* data = [YGORequest hotest];
    NSString* parsed = @"";
    if (![data isEqualToString:@""]) {
        parsed = [NSString stringWithUTF8String:parse([data UTF8String], 6)];
    }
    Hotest* result = [[Hotest alloc] init];
    @try {
        NSJSONSerialization* json = [NSJSONSerialization JSONObjectWithData:[parsed dataUsingEncoding:NSUTF8StringEncoding] options:NSJSONReadingMutableLeaves error:nil];
        if ([[json valueForKey:@"result"] intValue] == 0) {
            NSArray* arrSearch = [json valueForKey:@"search"];
            NSMutableArray<NSString*>* retSearch = [NSMutableArray array];
            for (int i = 0; i < arrSearch.count; i++) {
                [retSearch addObject:[arrSearch objectAtIndex:i]];
            }
            NSArray* arrCard = [json valueForKey:@"card"];
            NSMutableArray<HotCard*>* retCard = [NSMutableArray array];
            for (int i = 0; i < arrCard.count; i++) {
                HotCard* ci = [[HotCard alloc] init];
                ci.name = [YGOData replaceChars:[[arrCard objectAtIndex:i] valueForKey:@"name"]];
                ci.hashid = [[arrCard objectAtIndex:i] valueForKey:@"hashid"];
                [retCard addObject:ci];
            }
            NSArray* arrPack = [json valueForKey:@"pack"];
            NSMutableArray<HotPack*>* retPack = [NSMutableArray array];
            for (int i = 0; i < arrPack.count; i++) {
                HotPack* pi = [[HotPack alloc] init];
                pi.name = [YGOData replaceChars:[[arrPack objectAtIndex:i] valueForKey:@"name"]];
                pi.packid = [[arrPack objectAtIndex:i] valueForKey:@"packid"];
                [retPack addObject:pi];
            }
            result.search = retSearch;
            result.card = retCard;
            result.pack = retPack;
        }
    } @catch (NSException *exception) {
        
    } @finally {
        
    }
    
    return result;
}

@end
